File: leetcode/0002-add-two-numbers/addTwoNumbers.go
package leetcode

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var result = ListNode{}

	var cursor = &result
	var carry = 0
	for l1 != nil || l2 != nil {
		var num1 = 0
		var num2 = 0
		if l1 != nil {
			num1 = l1.Val
		}
		if l2 != nil {
			num2 = l2.Val
		}

		var sum = num1 + num2 + carry
		cursor.Next = &ListNode{Val: sum % 10}
		carry = sum / 10

		//next
		cursor = cursor.Next
		if l1 != nil {
			l1 = l1.Next
		}
		if l2 != nil {
			l2 = l2.Next
		}
	}

	if carry > 0 {
		cursor.Next = &ListNode{Val: carry}
	}

	return result.Next
}

File: leetcode/0002-add-two-numbers/addTwoNumbers.ts
import { ListNode } from "./helper";

function addTwoNumbers(
  l1: ListNode | null,
  l2: ListNode | null
): ListNode | null {
  const result = new ListNode();

  let cursor = result;
  let cursorl1 = l1;
  let cursorl2 = l2;

  let carry = 0;
  while (cursorl1 != null || cursorl2 != null) {
    const sum = (cursorl1?.val ?? 0) + (cursorl2?.val ?? 0) + carry;

    cursor.next = new ListNode(sum % 10);
    carry = Math.floor(sum / 10);
    console.log(sum % 10);

    // next
    cursor = cursor.next;
    cursorl1 = cursorl1?.next ?? null;
    cursorl2 = cursorl2?.next ?? null;
  }
  // 最后进位
  if (carry > 0) {
    cursor.next = new ListNode(carry);
  }

  return result.next;
}

export default addTwoNumbers;
